File: Machine Learning/taller_01_regresion_lineal/cxx/analytical_linear_regression.cxx

#include "CostLinearRegression.h"
#include <chrono>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>

int main( int argc, char** argv )
{
  using TScalar = long double;
  using TCost   = CostLinearRegression< TScalar >;

  if( argc < 2 )
  {
    std::cerr << "Usage: " << argv[ 0 ] << " filename" << std::endl;
    return( 1 );
  } // end if
  std::string filename = argv[ 1 ];

  // Read data
  auto sT = std::chrono::high_resolution_clock::now( );
  TCost c;
  std::ostringstream buf_str;
  std::ifstream input_file( filename.c_str( ), std::ios::binary );
  buf_str << input_file.rdbuf( );
  std::istringstream input( buf_str.str( ) );
  input_file.close( );
  input >> c;
  auto eT = std::chrono::high_resolution_clock::now( );
  double rT =
    std::chrono::duration_cast< std::chrono::nanoseconds >( eT - sT ).
    count( );
  rT *= 10e-10;
  std::cout
    << "Data read from " << filename << " in " << rT << " seconds " << std::endl;

  // Solve analytically
  TCost::TMatrix w;
  TCost::TScalar b;
  sT = std::chrono::high_resolution_clock::now( );
  c.analytic_solve( w, b );
  eT = std::chrono::high_resolution_clock::now( );
  rT =
    std::chrono::duration_cast< std::chrono::nanoseconds >( eT - sT ).
    count( );
  rT *= 10e-10;

  std::cout
    << "Analytical version done in " << rT << " seconds" << std::endl
    << "----------------------------" << std::endl
    << "---- Analytical results ----" << std::endl
    << "----------------------------" << std::endl
    << "w = " << w << std::endl
    << "b = " << b << std::endl
    << "J = " << c.evaluate( w, b ) << std::endl
    <<  "----------------------------" << std::endl;

  return( 0 );
}

// eof - analytical_linear_regression.cxx

File: Machine Learning/taller_01_regresion_lineal/python3/analytical_linear_regression.py
import sys, time
import numpy
from ReadData import ReadData
from CostLinearRegression import CostLinearRegression

# 1. Read data
data_path = 'p_04.txt'
sT = time.process_time( )
[ X_data, Y_data ] = ReadData(sys.argv[ 1 ])
eT = time.process_time( )
print( "Data read from", sys.argv[ 1 ], "in", float( eT - sT ), "seconds" )

# 2. Create numpy objects
sT = time.process_time( )
X = numpy.matrix( X_data )
Y = numpy.matrix( Y_data ).transpose( )
eT = time.process_time( )
print( "Data converted in", float( eT - sT ), "seconds" )

# 3. Analytical version
sT = time.process_time( )

linearRegression = CostLinearRegression( X, Y )
[w, b] = linearRegression.analytic_solve()
J = linearRegression.evaluate(w, b)
#[ w, b, J ] = AnalyticalLinearRegression( X, Y )
eT = time.process_time( )
print( "Analytical version done in", float( eT - sT ), "seconds" )
print( "----------------------------" )
print( "---- Analytical results ----" )
print( "----------------------------" )
print( "w =", w )
print( "b =", b )
print( "J =", J )
print( "----------------------------" )

## eof - analytical_linear_regression.py
